Lib/test/test_capi/test_codecs.py

```diff
@@ -516,5 +521,230 @@ def test_asrawunicodeescapestring(self):
+        self.assertEqual(decode(b, 'utf-8', NULL), s)
+        self.assertEqual(decode(b, 'latin1', 'strict'), b.decode('latin1'))
+        self.assertRaises(UnicodeDecodeError, decode, b, 'ascii', 'strict')
+        self.assertRaises(UnicodeDecodeError, decode, b, 'ascii', NULL)
+        self.assertEqual(decode(b, 'ascii', 'replace'), 'a' + '\ufffd'*9)
+        # todo: add more cases
+
+        # _codecs.decode only reports unknown errors policy when they are
+        # used (it has a fast path for empty bytes); this is different from
+        # PyUnicode_Decode which checks that both the encoding and the errors
+        # policy are recognized.
+        self.assertEqual(decode(b'', 'utf-8', 'unknown-errors-policy'), '')
+
+        self.assertRaises(TypeError, decode, NULL, 'ascii', 'strict')
+        # CRASHES decode(b, NULL, 'strict')
+
+    def test_codec_encoder(self):
+        with self.use_custom_encoder():
+            encoder = _testcapi.codec_encoder(self.encoding_name)
+            self.assertIs(encoder, self.codec_info.encode)
+
+    def test_codec_decoder(self):
+        with self.use_custom_encoder():
+            decoder = _testcapi.codec_decoder(self.encoding_name)
+            self.assertIs(decoder, self.codec_info.decode)
+
+    def test_codec_incremental_encoder(self):
+        with self.use_custom_encoder():
+            encoder = _testcapi.codec_incremental_encoder(self.encoding_name, 'strict')
+            self.assertIsInstance(encoder, self.codec_info.incrementalencoder)
+
+    def test_codec_incremental_decoder(self):
+        with self.use_custom_encoder():
+            decoder = _testcapi.codec_incremental_decoder(self.encoding_name, 'strict')
+            self.assertIsInstance(decoder, self.codec_info.incrementaldecoder)
+
+    def test_codec_stream_reader(self):
+        with self.use_custom_encoder():
+            encoding, stream = self.encoding_name, io.StringIO()
+            reader = _testcapi.codec_stream_reader(encoding, stream, 'strict')
+            self.assertIsInstance(reader, self.codec_info.streamreader)
+
+    def test_codec_stream_writer(self):
+        with self.use_custom_encoder():
+            encoding, stream = self.encoding_name, io.StringIO()
+            writer = _testcapi.codec_stream_writer(encoding, stream, 'strict')
+            self.assertIsInstance(writer, self.codec_info.streamwriter)
+
+class CAPICodecErrors(unittest.TestCase):
+
+    def setUp(self):
+        self.enterContext(import_helper.isolated_modules())
+        self.enterContext(import_helper.CleanImport('codecs'))
+        self.codecs = import_helper.import_module('codecs')
+
+    def test_codec_register_error(self):
+        self.assertRaises(LookupError, _testcapi.codec_lookup_error, 'custom')
+
+        def error_handler(exc):
+            raise exc
+
+        error_handler = mock.Mock(wraps=error_handler)
+        _testcapi.codec_register_error('custom', error_handler)
```

Comment:
Don't you have to unregister the error handler later?

Reply:
I would but I couldn't find an API for that :(

Reply:
See #124665.